### DISCUSSION TITLE
file:// not allowed to load local resource

### DISCUSSION THREAD

**Discussion by rafma0** [tauri-apps/tauri#1438] [Q&A] [answered]
hey, is there any way I can use `file://` to load pics / videos?

In electron it works after I do this
```js
  protocol.registerFileProtocol('file', (request, callback) => {
    const pathname = decodeURI(request.url.replace('file:///', ''));
    callback(pathname);
  });
```

I have a external config file where the user defines a local path and I need to show the pics inside, etc

this is one of the few reasons I'm still stuck with electron and I can't wait to migrate 😃 

thank you

**Comment by niemes** [comment, marked as answer]
Up, same problem. 

I would like to change the source of a videoPlayer and end up with this error/limitation. 
Any solution or documentation to load local file ?

Thank you

# EDIT
We should also mention that we need to add the :asset protocole in the csp for this to work normally.
**in the file tauri.conf.json add in the csp key "asset:"** 

I find the way to use local file, you need to get the link from the api like that : 
source : [Doc Tauri](https://tauri.studio/en/docs/api/js/modules/tauri)
````
      window.__TAURI__.dialog.open().then( (localPath) => {
        console.log("Video selected", localPath)
        
        let apiPath = window.__TAURI__.tauri.convertFileSrc(localPath)
        console.log('API Path', apiPath)
        window.videoPlayer.src = apiPath
        window.videoPlayer.play()
      })
````

have fun o/

**Reply by rafma0** [reply]
Your example is working for me, thank you 😃

One thing I noticed is Docs say you must omit the drive letter on Windows for convertFileSrc, but you need it to work

```js
window.__TAURI__.tauri.convertFileSrc('/a/b/c/d/e.mp4') -> 'https://asset.localhost//a/b/c/d/e.mp4' ❌
window.__TAURI__.tauri.convertFileSrc('c:\a\b\c\d\e.mp4') -> 'https://asset.localhost/c:a\bcde.mp4' ❌
window.__TAURI__.tauri.convertFileSrc('c:\\a\\b\\c\\d\\e.mp4') -> 'https://asset.localhost/c:\a\b\c\d\e.mp4' ✔️
window.__TAURI__.tauri.convertFileSrc('c:/a/b/c/d/e.mp4') -> 'https://asset.localhost/c:/a/b/c/d/e.mp4' ✔️
```

Another thing is it seems like videos with small file size (lower than 1mb maybe (?)) won't load, but they work fine on Edge. I'm investigating it a little more and opening a bug report later.

#2854

**Reply by niemes** [reply]
Im glad it works! 

On my side, since the last macOS update the asset: protocole dont work anymore. Blocked by CORS :/

**Reply by antoniotorres** [reply]
I have the same issue with Mac OS. Below you can see the CORS error that appears when testing the app.

![image](https://user-images.githubusercontent.com/4925102/143189661-4bcf8ac7-54c8-47c4-bde4-97bdd0e59d7f.png)

```bash
yarn tauri info
yarn run v1.22.10
warning package.json: No license field
warning ../package.json: No license field
$ tauri info
 app:spawn Running "/Users/antonio/tauri-app/node_modules/@tauri-apps/cli/bin/tauri-cli tauri info" +0ms


Operating System - Mac OS, version 12.0.1 X64

Node.js environment
  Node.js - 12.20.0
  @tauri-apps/cli - 1.0.0-beta.10
  @tauri-apps/api - 1.0.0-beta.8

Global packages
  npm - 6.14.8
  yarn - 1.22.10

Rust environment
  rustc - 1.53.0
  cargo - 1.53.0

App directory structure
/dist
/node_modules
/src-tauri
/src

App
  tauri.rs - 1.0.0-beta.8
  build-type - bundle
  CSP - default-src asset: blob: data: filesystem: ws: wss: http: https: tauri: 'unsafe-eval' 'unsafe-inline' 'self' img-src: 'self'
  distDir - ../dist
  devPath - http://localhost:3000
  framework - React

✨  Done in 5.47s.
```

**Reply by breadthe** [reply]
Thanks for this @niemes ! I was struggling with the same problem and completely overlooked [this section in the docs](https://tauri.app/v1/api/js/modules/tauri/#convertfilesrc).

In my case I'm dealing with images. Here's the [commit](https://github.com/breadthe/sd-buddy/commit/8ded008431f07f6a028ebcac2a73f10c76c193f4) where I applied this technique.

**Reply by mrdududu** [reply]
In  the documentation i found that "Additionally, asset must be added to [tauri.allowlist.protocol](https://tauri.app/v1/api/config/#allowlistconfig.protocol) in tauri.conf.json and its access scope must be defined on the assetScope array on the same protocol object."
But there is no any examples. How it must be looks in tauri.conf.json

**Reply by lighttigerXIV** [reply]
If someone was stuck like me with react i will leave this comment here.

I added this configs in my tauri.conf.json:
```json
"path": {
  "all": true
},
"fs": {
  "all": true,
  "scope": [
    "**"
  ]
},
"protocol": {
  "asset": true,
  "assetScope": ["**"]
}
```

And then in the file dialog i did this:

```javascript

//The imports you need
import { open } from "@tauri-apps/api/dialog"
import { convertFileSrc } from '@tauri-apps/api/tauri';

async function selectImage() {

  const selected = await open({
    multiple: false,
    filters: [{
      name: "Images",
      extensions: ["png", "webp", "jpg", "jpeg"]
    }]
  });

  if (selected !== null) {

    setSelectedImagePath(convertFileSrc(selected));
    console.log(selectedImagePath)
  }
}
```

Then just make something like this in your image:
```
<img src={selectedImagePath} className="h-200"></img>
```

I am impressed how i didn't found something so simple :3

**Reply by FabianLars** [reply]
by the way, if you're using tauri's file dialog you don't have to configure such a permissive scope. The via the dialog selected files will be added to the scope automatically :)

**Reply by abkarim** [reply]
The protocol section worked 
thank you so much 😊

**Reply by aaalloc** [reply]
This works for images, but beware that this currently doesn't work for videos and music files for Linux users  
https://github.com/tauri-apps/tauri/issues/3725

**Reply by kzhui125** [reply]
I have tested that asset protocol can't play large size video on macOS:  https://github.com/tauri-apps/tauri/issues/8589

**Reply by YMC-GitHub** [reply]
@lighttigerXIV Thank you! it works for me.

**Reply by mohammadmansour200** [reply]
@lighttigerXIV tried removing the all property from the fs config, and still works BTW; this is better as stated in the docs (security and bundle-size wise)
```json
 "fs": {
        "scope": ["**"]
      },
```

**Reply by GregJFlores** [reply]
> If someone was stuck like me with react i will leave this comment here.
> 
> I added this configs in my tauri.conf.json:
> 
> ```json
> "path": {
>   "all": true
> },
> "fs": {
>   "all": true,
>   "scope": [
>     "**"
>   ]
> },
> "protocol": {
>   "asset": true,
>   "assetScope": ["**"]
> }
> ```
> 
> And then in the file dialog i did this:
> 
> ```js
> //The imports you need
> import { open } from "@tauri-apps/api/dialog"
> import { convertFileSrc } from '@tauri-apps/api/tauri';
> 
> async function selectImage() {
> 
>   const selected = await open({
>     multiple: false,
>     filters: [{
>       name: "Images",
>       extensions: ["png", "webp", "jpg", "jpeg"]
>     }]
>   });
> 
>   if (selected !== null) {
> 
>     setSelectedImagePath(convertFileSrc(selected));
>     console.log(selectedImagePath)
>   }
> }
> ```
> 
> Then just make something like this in your image:
> 
> ```
> <img src={selectedImagePath} className="h-200"></img>
> ```
> 
> I am impressed how i didn't found something so simple :3

You are a life saver

**Reply by ShiroAky** [reply]
Thank you very much, I can now solve the problem in my application thanks to your explanation.

**Reply by alishefaee** [reply]
> > If someone was stuck like me with react i will leave this comment here.
> > I added this configs in my tauri.conf.json:
> > ```json
> > "path": {
> >   "all": true
> > },
> > "fs": {
> >   "all": true,
> >   "scope": [
> >     "**"
> >   ]
> > },
> > "protocol": {
> >   "asset": true,
> >   "assetScope": ["**"]
> > }
> > ```



Is this solution for version 1? How can we achieve this in v2?

**Reply by lighttigerXIV** [reply]
> > > If someone was stuck like me with react i will leave this comment here.
> > > I added this configs in my tauri.conf.json:
> > > ```json
> > > "path": {
> > >   "all": true
> > > },
> > > "fs": {
> > >   "all": true,
> > >   "scope": [
> > >     "**"
> > >   ]
> > > },
> > > "protocol": {
> > >   "asset": true,
> > >   "assetScope": ["**"]
> > > }
> > > ```
> 
> Is this solution for version 1? How can we achieve this in v2?

This works for me in v2. I had quite the trouble finding this out at the time XD:
```json
"app": {
  "security": {
    "csp": null,
    "assetProtocol": {
      "enable": true,
      "scope": {
        "requireLiteralLeadingDot": false,
        "allow": ["**/*"]
      }
    }
  }
}
```

**Reply by alishefaee** [reply]
From discord channel I got this help:
[https://tauri.app/reference/javascript/api/namespacecore/#convertfilesrc](https://tauri.app/reference/javascript/api/namespacecore/#convertfilesrc)
example:
```
<img
  src={convertFileSrc('F://element-v4//images//train-temp//19-12-06-29-6.jpg')}
  // src="file:///F:/element-v4/images/train-temp/19-12-06-29-6.jpg"
  alt="image didn't load"
  style={{
    height: "auto",
    width: "94%",
    borderStyle: "solid",
    borderWidth: "1px",
    borderColor: "grey",
  }}
/>
```
in ```setup``` method I should add my path to the scope:
```
            let conn = get_connection();
            let setting_service = SettingService::new(conn);
            match setting_service.find_one(&SettingEnum::BasePath) {
                Ok(setting_option) => {
                    if let Some(setting) = setting_option {
                        if let Some(path_str) = setting.value.as_str() {
                            let path_buf = PathBuf::from(path_str);
            
                            // Configure both scopes with the path
                            app.asset_protocol_scope()
                                .allow_directory(&path_buf, true)?;
                            app.fs_scope()
                                .allow_directory(&path_buf, true)?;
                        } else {
                            eprintln!("BasePath setting is not a string");
                        }
                    } else {
                        println!("No base path setting found in database");
                    }
                }
                Err(err) => {
                    // Handle database query error
                    eprintln!("Error querying base path setting: {}", err);
                }
            }
```

**Comment by xudong7** [comment]
This works in tauri2:
```
"security": {
  "csp": null,
  "assetProtocol": {
    "enable": true,
    "scope": ["**"]
  }
}
```
